src/Compilers/CSharp/Portable/Lowering/RuntimeAsyncIteratorRewriter/RuntimeAsyncIteratorRewriter.cs

```diff
@@ -0,0 +1,428 @@
+
+        var rewritter = new MoveNextAsyncRewriter(
+             F,
+             method,
+             stateField,
+             instanceIdField,
+             hoistedVariables,
+             nonReusableLocalProxies,
+             nonReusableFieldsForCleanup,
+             synthesizedLocalOrdinals,
+             stateMachineStateDebugInfoBuilder,
+             slotAllocatorOpt,
+             nextFreeHoistedLocalSlot,
+             diagnostics,
+             _currentField,
+             _disposeModeField,
+             _combinedTokensField);
+
+        BoundStatement runtimeAsyncMoveNextAsyncBody = rewritter.GenerateMoveNextAsync(body);
+
+        Debug.Assert(F.CurrentFunction is not null);
+        BoundStatement rewrittenBody = RuntimeAsyncRewriter.RewriteWithoutHoisting(runtimeAsyncMoveNextAsyncBody, F.CurrentFunction, F.CompilationState, F.Diagnostics);
+
+        F.CloseMethod(rewrittenBody);
+    }
+
+    private MethodSymbol GetMoveNextAsyncMethod()
+    {
+        Debug.Assert(_currentField is not null);
+
+        MethodSymbol IAsyncEnumerator_MoveNextAsync = F.WellKnownMethod(WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__MoveNextAsync);
+        var IAsyncEnumerator = IAsyncEnumerator_MoveNextAsync.ContainingType;
+        var IAsyncEnumeratorOfElementType = IAsyncEnumerator.Construct(_currentField.Type);
+        MethodSymbol IAsyncEnumeratorOfElementType_MoveNextAsync = IAsyncEnumerator_MoveNextAsync.AsMember(IAsyncEnumeratorOfElementType);
+        return IAsyncEnumeratorOfElementType_MoveNextAsync;
+    }
+
+    /// <summary>
+    /// Generates the Current property.
+    /// </summary>
+    private void GenerateIAsyncEnumeratorImplementation_Current()
+    {
+        AsyncRewriter.AsyncIteratorRewriter.GenerateIAsyncEnumeratorImplementation_Current(_currentField, this, F);
+    }
+
+    /// <summary>
+    /// Generates the `ValueTask IAsyncDisposable.DisposeAsync()` method as a runtime-async method.
+    /// </summary>
+    [SuppressMessage("Style", """VSTHRD200:Use "Async" suffix for async methods""", Justification = "Standard naming convention for generating 'IAsyncDisposable.DisposeAsync'")]
+    private void GenerateIAsyncDisposable_DisposeAsync()
+    {
+        // Produce:
+        //  if (state >= StateMachineStates.NotStartedStateMachine /* -1 */)
+        //  {
+        //      throw new NotSupportedException();
+        //  }
+        //  if (state == StateMachineStates.FinishedStateMachine /* -2 */)
+        //  {
+        //      return;
+        //  }
+        //  disposeMode = true;
+        //  AsyncHelpers.Await(MoveNextAsync());
+        //  return;
+
+        Debug.Assert(_currentField is not null);
+        Debug.Assert(stateField is not null);
+        Debug.Assert(_disposeModeField is not null);
+        MethodSymbol IAsyncDisposable_DisposeAsync = F.WellKnownMethod(WellKnownMember.System_IAsyncDisposable__DisposeAsync);
```

Comment:
Please assert that the return type of this method is `ValueTask` from the BCL, not some other source.